src/legacy/core_plugins/kibana/public/visualize/wizard/new_vis_modal.tsx

```diff
@@ -45,33 +53,65 @@ class NewVisModal extends React.Component<TypeSelectionProps> {
   constructor(props: TypeSelectionProps) {
     super(props);
     this.isLabsEnabled = chrome.getUiSettingsClient().get('visualize:enableLabs');
+
+    this.state = {
+      showSearchVisModal: false,
+    };
   }
 
   public render() {
     if (!this.props.isOpen) {
       return null;
     }
 
-    return (
-      <EuiOverlayMask>
-        <EuiModal onClose={this.props.onClose} maxWidth={'100vw'} className="visNewVisDialog">
+    const selectionModal =
+      this.state.showSearchVisModal && this.state.visType ? (
+        <EuiModal onClose={this.onCloseModal} className="visNewVisSearchDialog">
+          <SearchSelection onSearchSelected={this.onSearchSelected} visType={this.state.visType} />
+        </EuiModal>
+      ) : (
+        <EuiModal onClose={this.onCloseModal} maxWidth={'100vw'} className="visNewVisDialog">
           <TypeSelection
             showExperimental={this.isLabsEnabled}
             onVisTypeSelected={this.onVisTypeSelected}
             visTypesRegistry={this.props.visTypesRegistry}
           />
         </EuiModal>
-      </EuiOverlayMask>
-    );
+      );
+
+    return <EuiOverlayMask>{selectionModal}</EuiOverlayMask>;
   }
 
+  private onCloseModal = () => {
+    this.setState({ showSearchVisModal: false });
+    this.props.onClose();
+  };
+
   private onVisTypeSelected = (visType: VisType) => {
-    const baseUrl =
-      visType.requiresSearch && visType.options.showIndexSelection
-        ? `#${VisualizeConstants.WIZARD_STEP_2_PAGE_PATH}?`
-        : `#${VisualizeConstants.CREATE_PATH}?`;
-    const params = [`type=${encodeURIComponent(visType.name)}`, ...this.props.editorParams!];
+    if (visType.requiresSearch && visType.options.showIndexSelection) {
+      this.setState({
+        showSearchVisModal: true,
+        visType,
+      });
+    } else {
+      const params = [`type=${encodeURIComponent(visType.name)}`, ...this.props.editorParams!];
+      this.props.onClose();
+      location.assign(`${baseUrl}${params.join('&')}`);
+    }
+  };
+
+  private onSearchSelected = (searchId: string, searchType: string) => {
     this.props.onClose();
+
+    if (!this.state.visType) {
```

Comment:
This check shouldn't be necessary - line 69 only renders the SearchSelection if `this.state.visType` is set. Am I missing something?

Reply:
Since `visType` is optional (it's defined after vis type selection by the user), `tslint` warns at line [107](https://github.com/elastic/kibana/pull/30196/files#diff-a52663fb9e8ccb0ee2abd6f8c8cded68R107). 
Since `onSearchSelected` will be invoked when `visType` is already defined, I removed this check and used a non-null assertion at 107 line